### PR TITLE
Show sharees via propfind

### PR DESCRIPTION
**Pull request by tobiasKaminsky**
:exclamation: TODO: once this is merged, change/check OCFileListAdapter: showShareAvatar boolean :exclamation: 

Steps to test
- create a file "share"
- share this file to userA and userB
- do a propfind with <nc:sharees/>, e.g.:
```
curl --request PROPFIND \
  --url http://localhost/nc/remote.php/webdav/share \
  --header 'authorization: Basic dG9iaTplOGVrei1SbVpaVC1DMkdDTi01clk2My1BeTl5QQ==' \
  --header 'content-type: application/xml' \
  --header 'ocs-apirequest: true' \
  --cookie 'ocjycgrudn78=vgovrq3bgraie6rs6ai2sl42ag; oc_sessionPassphrase=oyqxSU8b6n4N8Aj0yfObGeheppYvsdzgMhEkMgHw76nWBRLVRRM%252FZcjxqQ5QHtm3knScFxIKIjCzDasJYhLEYxRsvabseMjOqDSa%252BNSkI%252Fi%252FDrm6cKlWmA4QBr3NtcGI; nc_sameSiteCookielax=true; nc_sameSiteCookiestrict=true' \
  --data '<?xml version="1.0" encoding="UTF-8"?>
<D:propfind xmlns:D="DAV:">
<D:prop>
	
<sharees xmlns="http://nextcloud.org/ns" />
</D:prop></D:propfind>
'
```
- see: 
	- Federated sharee: `<nc:sharees>tobi@http://localhost/nc15</nc:sharees>`
	- Multiple sharees on same server: `<nc:sharees>admin, Tobi Kaminsky</nc:sharees>`
	- One sharee on same server: `<nc:sharees>admin</nc:sharees>`

Please note that this shows only the user you have created & shared a file to.
The other way around (receiving a share) is already exists via < owner-id> and < owner-display-name>.

Signed-off-by: tobiasKaminsky <tobias@kaminsky.me>

### CODE REVIEW

File: apps/dav/lib/Connector/Sabre/Node.php [hunk 1]
```diff
@@ -290,6 +293,64 @@ public function getSharePermissions($user) {
...
+		}
+		$types = [
+			Share::SHARE_TYPE_USER,
+			Share::SHARE_TYPE_REMOTE, 
+			Share::SHARE_TYPE_GROUP,
+			Share::SHARE_TYPE_CIRCLE,
+			Share::SHARE_TYPE_ROOM
+		];
+
+		if ($this->getPath() === "/") {
```
Comment (blizzz) [inline apps/dav/lib/Connector/Sabre/Node.php]: or-connect it with the condition in line 303

File: apps/dav/lib/Connector/Sabre/Node.php [hunk 2]
```diff
@@ -290,6 +293,64 @@ public function getSharePermissions($user) {
...
+
+		foreach ($types as $shareType) {
+			$shares = $this->shareManager->getSharesBy($user, $shareType, $path, false, -1, 0);
+			foreach ($shares as $share) {
+				if ($share->getSharedBy() === $user) {
+					$sharees[] = $share->getSharedWith();
+				}
+			}
+		}
+		return implode(', ', $sharees);
```
Comment (blizzz) [inline apps/dav/lib/Connector/Sabre/Node.php]: ain't it better return the array and format it where the output is being handled (the callee)? This also only works as long as "," is prohibited as identifier. Further, you have all the types mixed in and identifiers are not unique across those types. → What do you want to achieve eventually?

**Review by blizzz — COMMENTED** [review]
too ambiguous maybe

File: apps/dav/lib/Connector/Sabre/Node.php (hunk 1 above)
Comment (tobiasKaminsky) [reply apps/dav/lib/Connector/Sabre/Node.php]: :+1:

File: apps/dav/lib/Connector/Sabre/Node.php (hunk 2 above)
Comment (tobiasKaminsky) [reply apps/dav/lib/Connector/Sabre/Node.php]: If I simply return $sharees, I do get that as plain string, e.g. "adminuser1user2".
If you have a better way, please feel free to change it or give me a hint.
I do not "care" as long as the client has a list of userIds of all sharees.

Comment (tobiasKaminsky) [reply apps/dav/lib/Connector/Sabre/Node.php]: I removed Circle and Rooms, so that now only user, remote user and groups are returned.

**Comment by blizzz** [comment]
as discussed on irc, we agreed about not returning grouped by type and and each user, group etc. is inside an id-tag within the appropriate element.

File: apps/dav/lib/Connector/Sabre/Node.php [hunk 3]
```diff
@@ -290,6 +293,62 @@ public function getSharePermissions($user) {
...
+
+		$path = $this->getPath();
+
+		if ($path !== null) {
+			$userFolder = \OC::$server->getRootFolder()->getUserFolder($user);
+			try {
+				$path = $userFolder->get($path);
+				$this->lock($path);
+			} catch (\OCP\Files\NotFoundException $e) {
+				throw new OCSNotFoundException($this->l->t('Wrong path, file/folder doesn\'t exist'));
```
Comment (rullzer) [inline apps/dav/lib/Connector/Sabre/Node.php]: This should not be here

File: apps/dav/lib/Connector/Sabre/Node.php [hunk 4]
```diff
@@ -290,6 +293,62 @@ public function getSharePermissions($user) {
...
+			} catch (LockedException $e) {
+				throw new OCSNotFoundException($this->l->t('Could not lock path'));
```
Comment (rullzer) [inline apps/dav/lib/Connector/Sabre/Node.php]: Same. This is not an OCS endpoint

**Review by rullzer — CHANGES_REQUESTED** [review]
So I knew this looked familiar.

Because the way this is implemented it won't scale. If you request this property for a user on a share with 200 entries. It will fire off 300 SQL requests.

We should bundle it with https://github.com/nextcloud/server/blob/master/apps/dav/lib/Connector/Sabre/SharesPlugin.php

**Comment by tobiasKaminsky** [comment]
Bump, can anyone help me here? Or direct me in the right direction? :-)

**Comment by tobiasKaminsky** [comment]
I really would like to have this in.

**Comment by rullzer** [comment]
Let me fix the tests this evening

**Comment by rullzer** [comment]
Ok so the test require some more work as they were/are not the cleanest. I'll have a look tomorrow.

**Comment by rullzer** [comment]
Test all look good. Time for reviews.

**Review by rullzer — APPROVED** [review]
All good from my part

**Comment by rullzer** [comment]
Small note. Once we have this we could also enhance the server UI.
We could show who a file is shared to more easily etc.

File: apps/dav/lib/Connector/Sabre/ShareeList.php [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	/**
+	 * The xmlSerialize metod is called during xml writing.
+	 *
+	 * @param Writer $writer
+	 * @return void
+	 */
+	function xmlSerialize(Writer $writer) {
+		foreach ($this->shares as $share) {
+			$writer->startElement('{' . self::NS_NEXTCLOUD . '}sharee');
+			$writer->writeElement("id", $share->getSharedWith());
```
Comment (georgehrke) [inline apps/dav/lib/Connector/Sabre/ShareeList.php]: The inner Elements id, display name, and type should be namespaces as well

File: apps/dav/lib/Connector/Sabre/ShareeList.php [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+			$writer->writeElement("displayName", $share->getSharedWithDisplayName());
```
Comment (georgehrke) [inline apps/dav/lib/Connector/Sabre/ShareeList.php]: You usually don't use camelCase in XML. So either displayname or display-name please :)

**Review by georgehrke — CHANGES_REQUESTED** [review]

**Comment by georgehrke** [comment]
Besides those two remarks it looks good and works

**Comment by tobiasKaminsky** [comment]
Can you test this please with an admin account:
```
<nc:sharees>
          <nc:sharee>
            <nc:id>admin</nc:id>
            <nc:display-name/>
            <nc:type>1</nc:type>
          </nc:sharee>
          <nc:sharee>
            <nc:id>user1</nc:id>
            <nc:display-name>User One</nc:display-name>
            <nc:type>0</nc:type>
          </nc:sharee>
          <nc:sharee>
            <nc:id>user2</nc:id>
            <nc:display-name>user2</nc:display-name>
            <nc:type>0</nc:type>
          </nc:sharee>
        </nc:sharees>
``` 

For some reason the display name of an admin is missing, but it should be at least userId.
It is working for other users…
I just want to be sure if this is a problem of my local system…

File: apps/dav/lib/Connector/Sabre/ShareeList.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+	 * The xmlSerialize metod is called during xml writing.
+	 *
+	 * @param Writer $writer
+	 * @return void
+	 */
+	function xmlSerialize(Writer $writer) {
+		foreach ($this->shares as $share) {
+			$writer->startElement('{' . self::NS_NEXTCLOUD . '}sharee');
+			$writer->writeElement('{' . self::NS_NEXTCLOUD . '}id', $share->getSharedWith());
+			$writer->writeElement('{' . self::NS_NEXTCLOUD . '}display-name', $share->getSharedWithDisplayName());
```
Comment (georgehrke) [inline apps/dav/lib/Connector/Sabre/ShareeList.php]:
```suggestion
			$writer->writeElement('{' . self::NS_NEXTCLOUD . '}display-name', $share->getSharedWithDisplayName() ?? $share->getSharedWith());
```

Comment (tobiasKaminsky) [reply apps/dav/lib/Connector/Sabre/ShareeList.php]: I wrote a test for android library and there the problem also exists.
It seems that always one of the display names is set to userId.

**Comment by tobiasKaminsky** [comment]
BUG:

- create folder "testFolder"
- share it to user1, user2
- query "/remote.php/webdav/", see 
```
 <d:response>
    <d:href>/remote.php/webdav/testFolder/</d:href>
    <d:propstat>
      <d:prop>
        <nc:sharees>
          <nc:sharee>
            <nc:id>user1</nc:id>
            <nc:display-name>User One</nc:display-name>
            <nc:type>0</nc:type>
          </nc:sharee>
          <nc:sharee>
            <nc:id>user2</nc:id>
            <nc:display-name>user2</nc:display-name>
            <nc:type>0</nc:type>
          </nc:sharee>
        </nc:sharees>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
```

- query "/remote.php/webdav/testFolder", see:

```
 <d:response>
    <d:href>/remote.php/webdav/testFolder/</d:href>
    <d:propstat>
      <d:prop>
        <nc:sharees>
          <nc:sharee>
            <nc:id>user1</nc:id>
            <nc:display-name>User One</nc:display-name>
            <nc:type>0</nc:type>
          </nc:sharee>
        </nc:sharees>
      </d:prop>
      <d:status>HTTP/1.1 200 OK</d:status>
    </d:propstat>
  </d:response>
```

**Comment by tobiasKaminsky** [comment]
Bug fixed via latest commit.

**Comment by tobiasKaminsky** [comment]
From my side this is ready to get in

**Comment by rullzer** [comment]
Fixed the failing tests

**Review by georgehrke — APPROVED** [review]
Seems to do what its supposed do and webdav compliance wise I'm also happy with it 👍

**Pull request merged**